modules/dataframe/src/util/dataframeutil.cpp

```diff
@@ -0,0 +1,325 @@
+                                    keyColumn, indexCol1->getBuffer()->getDataFormat()->getString(),
+                                    indexCol2->getBuffer()->getDataFormat()->getString()),
+                        IVW_CONTEXT_CUSTOM("dataframeutil::innerJoin"));
+    }
+
+    // locate matching rows and determine row indices
+    auto rows = indexCol1->getBuffer()
+                    ->getRepresentation<BufferRAM>()
+                    ->dispatch<std::pair<std::vector<size_t>, std::vector<size_t>>>(
+                        [rightBuffer = indexCol2->getBuffer()](auto typedBuf) {
+                            using ValueType = util::PrecisionValueType<decltype(typedBuf)>;
+
+                            const auto& left = typedBuf->getDataContainer();
+                            const auto& right = static_cast<const BufferRAMPrecision<ValueType>*>(
+                                                    rightBuffer->getRepresentation<BufferRAM>())
+                                                    ->getDataContainer();
+
+                            std::vector<size_t> rowsLeft;
+                            std::vector<size_t> rowsRight;
+                            for (auto&& [i, key] : util::enumerate(left)) {
+                                if (auto it = util::find(right, key); it != right.end()) {
+                                    rowsLeft.push_back(i);
+                                    rowsRight.push_back(std::distance(right.begin(), it));
+                                }
+                            }
+                            return std::make_pair(rowsLeft, rowsRight);
+                        });
+
+    IVW_ASSERT(rowsLeft.size() == rowsRight.size(), "incorrect number of matching row indices");
+
+    auto dataframe = std::make_shared<DataFrame>();
+
+    auto addColumns = [&dataframe, keyColumn](const DataFrame& srcDataFrame,
+                                              const std::vector<size_t>& rows, bool skipKeyCol) {
+        for (auto srcCol : srcDataFrame) {
+            if (srcCol == srcDataFrame.getIndexColumn()) continue;
+            if (skipKeyCol && (srcCol->getHeader() == keyColumn)) {
+                continue;
+            }
+
+            srcCol->getBuffer()->getRepresentation<BufferRAM>()->dispatch<void>(
+                [&dataframe, header = srcCol->getHeader(), rows](auto typedBuf) {
+                    auto dst = util::transform(
+                        rows, [src = typedBuf->getDataContainer()](size_t i) { return src[i]; });
+                    dataframe->addColumn(header, dst);
+                });
+        }
+    };
+
+    addColumns(left, rows.first, false);
+    addColumns(right, rows.second, true);
+
+    dataframe->updateIndexBuffer();
+    return dataframe;
+}
+
+std::shared_ptr<DataFrame> combineDataFrames(std::vector<std::shared_ptr<DataFrame>> dataFrames,
+                                             bool skipIndexColumn, std::string skipcol) {
```

Comment:
Simply combine(...) because it is in the namespace?

Reply:
that was pre-existing prior this PR (similar to `cloneBufferRange`)